src/Components/InfiniteScroll.tsx

```diff
@@ -55,37 +69,140 @@ const DisplayPreviousPage = compose<IDisplayPreviousPageProps<any>, IDisplayPrev
+    pageHeightMap,
+    previousPageHeightRef,
+    recordPageHeight,
+  }) => ({
+    updatePageHeightMap: (query) => {
+      const currentPageHeight = currentPageHeightRef.current.getBoundingClientRect().height;
+      let previousPageHeight;
+      let nextPageHeight;
+
+      const updatedPageHeightMap = {
+        ...pageHeightMap,
+        [query.response.currentPage]: currentPageHeight,
+      };
+
+      if (query.hasPreviousPage()) {
+        previousPageHeight = previousPageHeightRef.current.getBoundingClientRect().height;
+        updatedPageHeightMap[query.response.previousPage] = previousPageHeight;
+      }
+
+      if (query.hasNextPage()) {
+        nextPageHeight = nextPageHeightRef.current.getBoundingClientRect().height;
+        updatedPageHeightMap[query.response.nextPage] = nextPageHeight;
+      }
+
+      recordPageHeight(updatedPageHeightMap);
+      return updatedPageHeightMap;
+    },
+  })),
   withStateHandlers<{ lastScrollTop: number }, { handleScroll }, IInfiniteScrollInternalProps<any>>(
     { lastScrollTop: 0 },
     {
-      handleScroll: ({ lastScrollTop }, { query, updateQuery }) => (clientHeight: number, scrollHeight: number, currentScrollTop: number) => {
-        const scrollingDown = currentScrollTop > lastScrollTop;
-        const scrollBottom = scrollHeight - currentScrollTop - clientHeight;
+      handleScroll: ({ lastScrollTop }, { disableVirtualScrolling, estimatedPageHeight, query, updatePageHeightMap, updateQuery }) =>
+        (clientHeight: number, scrollHeight: number, currentScrollTop: number) => {
+          const scrollingDown = currentScrollTop > lastScrollTop;
+          const currentScrollBottom = scrollHeight - currentScrollTop - clientHeight;
 
-        if (scrollingDown && scrollBottom < (clientHeight / 2) && query.hasNextPage()) {
-          updateQuery(query.getNextPage());
-        }
+          if (!disableVirtualScrolling) {
+            const updatedPageHeightMap = updatePageHeightMap(query);
 
-        return { lastScrollTop: currentScrollTop };
-      },
+            let nextPageToLoad = 1;
+            let nextPageScrollTop = updatedPageHeightMap[1];
+
+            while (nextPageScrollTop < currentScrollTop && nextPageToLoad < query.response.totalPages) {
+              nextPageToLoad++;
+              nextPageScrollTop += nextPageToLoad in updatedPageHeightMap
+                ? updatedPageHeightMap[nextPageToLoad]
+                : estimatedPageHeight;
+            }
+
+            if (query.response.currentPage !== nextPageToLoad) {
+              updateQuery(query.query.page(nextPageToLoad));
+            }
+          } else if (scrollingDown && currentScrollBottom < (clientHeight / 2) && query.hasNextPage()) {
+            updateQuery(query.getNextPage());
+          }
+
+          return { lastScrollTop: currentScrollTop };
+        },
     },
   ),
-  withProps(({ handleScroll, debounceTime }) => ({
+  withProps(({ debounceTime, handleScroll }) => ({
```

Comment:
I think we should use `withPropsOnChange` instead of `withProps` so we do not redeclare `handleScrollDebounced` whenever any of the props have changed.

Reply:
changed the first to withPropsOnChange and the second following to withHandlers